### DISCUSSION TITLE
fetch() SSL disable certificate validation

### DISCUSSION THREAD

**Discussion by marek-benes** [nodejs/help#4636] [Q&A] [answered]
Hello, is there any way to disable SSL certificate validation for self-signed certificates or expired ones?

Thank you.

**Comment by bnoordhuis** [comment]
Yes, pass `{ rejectUnauthorized: false }` in the options argument to `tls.connect()`, `https.request()`, etc.

**Reply by marek-benes** [reply]
`        const response = await fetch("https://google.com", {
            method: "get",
            body: body
        });
`

But there is no option argument to pass`rejectUnauthorized` property.

**Reply by bnoordhuis** [reply]
Oh sorry, you did mention fetch(). That's a web API so no, no escape hatch. You wouldn't have one in the browser either.

**Reply by shivkumar-jirwankar** [reply]
any solution for fetch api?

**Comment by gusttavonl** [comment]
use this in your fetch

```js
const https = require('https'); // or import https from 'https'
const agent = new https.Agent({
    rejectUnauthorized: false,
});
fetch("https://google.com", { method: "get", body: body, agent });
```

**Reply by Ju777** [reply]
Hello @GustavoNoronha0 , 
what you wrote looks exactly what I do need, thanks about it.

I got this error message while compiling the line `` import https from 'https' `` (in React) :
```
Module not found: Error: Can't resolve 'https' in '/my_gophish_directory/'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "https": require.resolve("https-browserify") }'
	- install 'https-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "https": false }
```

Do you have any idea about the reason ?

Have a nice day.
Julien

**Reply by gusttavonl** [reply]
Hello @Ju777, you need to install the http module in your project. Run this in terminal `npm install https`. Tell me if it worked, have a nice day too

**Reply by Ju777** [reply]
Thanks @GustavoNoronha0 ,

actually I just tried before you wrote. And the result is the same (same error message).

**Reply by gusttavonl** [reply]
Sorry @Ju777, then try to do what the error says, in your `webpack.config.js`
```
fallback: {
    "https-browserify": require.resolve('https-browserify'),
  } 
```

**Reply by Ju777** [reply]
No problem :) @GustavoNoronha0 .
The problem is that I can't see any webpack configuration file at the root of my application. 
As far as I know Webpack is included with any React app right ?

**Reply by gusttavonl** [reply]
Right @Ju777, I think a thread answered about this issue for you [To Fix The Problem](https://github.com/facebook/create-react-app/issues/11756#issuecomment-1001162736)

**Reply by Ju777** [reply]
Ok @GustavoNoronha0 , I am having a look at it. Thanks.

**Reply by Ju777** [reply]
@GustavoNoronha0 may I ask what are your React version, Node version, and Webpack version ?

**Reply by gusttavonl** [reply]
@Ju777 Of course, React Version: 18.2.0, Node: 16.16.0 and Webpack: 5.75.0

**Reply by Ju777** [reply]
Ok thanks @GustavoNoronha0 .

There is a weird thing, because I did not install any Webpack version as I create my app with CRA ``npx create-react-app my_app_name`` . And as far as I know, there is a Webpack "under the hood" in that case.

I think that's the reason why I don't have any ``webpack.config.js`` .

**Reply by juinson** [reply]
so, do you guys have any idea to solve this case ? 
coz i spend 2 times 24 hours and had no idea

**Reply by juinson** [reply]
i have tried https-proxy-agent and so on, even use server proxy

**Reply by Taschenbuch** [reply]
this whole thread in this discussion is confusing. 
Are we talking about the same thing: build in fetch in nodejs?
The nodejs build in fetch does not accept a "agent" parameter. sounds like you talk about node-fetch package or browser api fetch. https is part of nodejs so no need to install it.
But the other accepted undici answer works for me.

**Comment by exseniarch** [comment]
Obviously on localhost it's a matter of CORS

http://www.linkedin.com/in/exseniarch

**Comment by KhafraDev** [comment, marked as answer]
Using the built in fetch with undici, you can do

```mjs
import { Agent, setGlobalDispatcher } from 'undici'

const agent = new Agent({
  connect: {
    rejectUnauthorized: false
  }
})

setGlobalDispatcher(agent)

await fetch('...')
```

**Reply by dev1-akkess** [reply]
I'm confused by this.  My understanding is that the built-in fetch added in v16.15.0 was built on top of undici, but there doesn't seem to be any undici to import.  Do I have to install the npm package for it and this will overwrite or augment the built-in one, or do I not understand how Node works?

**Reply by KhafraDev** [reply]
Yes, you have to install undici. `setGlobalDispatcher` works by using a global symbol so the global dispatcher can be set from undici but also used in node's fetch.

**Reply by dev1-akkess** [reply]
Okay makes sense, thank you!

**Reply by mherb63** [reply]
when I do the import
`import { Agent, setGlobalDispatcher } from 'undici'`

I get this error in the console
<img width="737" alt="image" src="https://github.com/nodejs/node/assets/1205814/a3d6c64f-be56-4fe1-81af-0b2b2044576e">


any ideas?

**Reply by KhafraDev** [reply]
undici doesn't work in the browser

**Reply by mherb63** [reply]
oh I see...thanks for the reply

**Comment by silverwind** [comment]
For anyone looking to do this without dependencies, e.g. for scripting use:

```js
#!/usr/bin/env -S node --no-warnings
process.env.NODE_TLS_REJECT_UNAUTHORIZED = 0;
await fetch("https://localhost");
```

The lengths one has to go through with built-in `fetch` for the simplest of things are truly astonishing.

**Reply by LqdBcnAtWork** [reply]
Why can't we set this per-fetch call?

Would be so much better than throwing a warning and disabling tls basically globally. Like really, we're not writing code to be run client side when it comes to Node. Let me send post requests to the modem I'm trying to auto configure that has a self-signed SSL.

I'm not expecting answers, just frustrated at how long it took me to find an actual answer to this.

Edit: very important word forgotten

**Reply by guscsales** [reply]
This works, but anyone that reading this message, please use only for development mode
